[incubator/logstash] Revamp logstash from Deployment to StatefulSet to support Persisted Queues

josdotso opened the pull request:
* Give boilerplate for kafka input and output.
* Prepare for persisted queues use case by default.
* Configure logstash via env var instead of configmap.
* Prepare for real HTTP ingress.
* Disable legacy syslog protocols by default, keep boilerplate.
* Nominate `beats` as the only default input.
* Support exposing non-HTTP TCP ports via separate LoadBalancer service.
* Parameterize many prod-centric settings.
* Support easy Prometheus scraping of logstash_exporter running in each pod.
  * https://github.com/BonnierNews/logstash_exporter
* Upgrade to logstash-oss:6.2.2
* Use logstash health API for logstash health checks
* Optionally seeds data at Prometheus for this awesome Logstash Grafana dashboard
  * https://grafana.com/dashboards/2525
* Set default pipeline to beats->ES
* Include official boilerplate for ES output (when beats input)
* Describe multi-pipeline usage recommendations.
* Show NOTES.txt for HTTP only when applicable.
* Reverse erroneous addition of josdotso as maintainer.
  * (Would like to revisit in the future, once I become more experienced with the repo and chart.)
  * If you are actively seeking maintainers, please let me know. I work with some other folks who may wish to work toward that goal with me.
* Add notes on best practice of loading beats-generated.
  index-template into elasticsearch.
* Refactored AWS-specific annotations as examples at values file.
  * Give example of service annotation to realize AWS Route53 DNS record for internal (or otherwise) ELB.
  * Give example of how to make serviceLoadBalancer an internal AWS ELB.
* Implemented storageClass best practices at values and statefulset.

Commit:
[incubator/logstash] Revamp logstash as scraped StatefulSet

* Give boilerplate for kafka input and output.
* Default to no functional dependency on elasticsearch.
* Prepare for persisted queues use case by default.
* Configure logstash via env var instead of configmap.
* Prepare for real HTTP ingress.
* Disable legacy syslog protocols by default, keep boilerplate.
* Nominate `beats` as the only non-HTTP default input.
* Support exposing non-HTTP TCP ports via internal-to-VPC ELB.
* Parameterize many prod-centric settings.
* Support easy Prometheus scraping of logstash_exporter running in each pod.
  * https://github.com/BonnierNews/logstash_exporter
* Upgrade to logstash-oss:6.2.2
* Use logstash health API for logstash health checks

ghost commented:
/assign @mattfarina

Commit:
[incubator/logstash] Fix missing backquotes in README

ghost commented:
<img width="705" alt="screen shot 2018-02-23 at 10 28 45 am" src="https://user-images.githubusercontent.com/13571670/36601900-671187e2-1884-11e8-91e9-a79ee76e0229.png">

^ chaoskube vs. Kubernetes rescheduling a logstash pod. view is from aforementioned Grafana dashboard.

Commit:
[incubator/logstash] Apply more logstash best practices

- Set default pipeline to beats->ES
- Include official boilerplate for ES output (when beats input)
- Give example of private zone dns record for internal ELB.
- Disable HTTP input by default, RE: beats as default input.
- Describe multi-pipeline usage recommendations.
- Show NOTES.txt for HTTP only when applicable.

unguiculus commented:
cc @christian-roggia

unguiculus commented on incubator/logstash/Chart.yaml:
```diff
@@ -15,3 +15,5 @@ maintainers:
   email: jrodgers@powerhrg.com
 - name: christian-roggia
   email: christian.roggia@gmail.com
+- name: josdotso
```
Please clarify why you added yourself as maintainer. You will be expected to continue contributing and review PRs.

@christian-roggia Are you fine with this?

unguiculus commented on incubator/logstash/templates/service-aws-lb-internal.yaml:
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.serviceAwsLoadBalancerInternal.enabled }}
```
The service should not be AWS-specific.

unguiculus commented on incubator/logstash/templates/statefulset.yaml:
```diff
@@ -0,0 +1,142 @@
+        name: data
+      spec:
+        accessModes: ["ReadWriteOnce"]
+        storageClassName: {{ .Values.persistentVolumes.storageClass }}
```
Please follow our best practices for storage class configuration.

See #1869.

unguiculus commented on incubator/logstash/templates/statefulset.yaml:
```diff
@@ -0,0 +1,142 @@
+      annotations:
+        checksum/patterns: {{ include (print $.Template.BasePath "/patterns-config.yaml") . | sha256sum }}
+        checksum/pipeline: {{ include (print $.Template.BasePath "/pipeline-config.yaml") . | sha256sum }}
+      {{- if .Values.exporter.logstash.enabled }}
```
Please make these annotations configurable separately. You don't need them if you use the prometheus-operator.

unguiculus reviewed (changes requested):
Thank you for this excellent contribution. I especially appreciate updating the chart to currrent best practices.

Commit:
[incubator/logstash] Refining PR #3844 per review comments

- Reverse erroneous addition of josdotso as maintainer.
- Add notes on best practice of loading beats-generated.
  index-template into elasticsearch.
- Refactored AWS-specific annotations as examples at values file.
- Reworked `serviceAwsLoadBalancerInternal` as agnostic `serviceLoadBalancer`.
- Simplified name of serviceLoadBalancer template and resource.
- Implemented storageClass best practices at values and statefulset.

ghost commented:
Thanks again, @unguiculus!  I updated the top comment and responded to comments inline after pushing new commits.

Here's a view of Grafana Logstash Dashboard showing Beats events coming through and going out to Elasticsearch.
![screen shot 2018-02-23 at 11 15 47 pm](https://user-images.githubusercontent.com/13571670/36625546-8b528f5a-18ef-11e8-90ba-9f97d1bda0f8.png)

ghost commented:
FYI: I had to amend my last commit as my local build of the chart slipped past me and into the commit.  Sorry about that!

Commit:
[incubator/logstash] Fix minor typos at README

unguiculus commented on incubator/logstash/templates/service-lb.yaml:
```diff
@@ -0,0 +1,27 @@
+{{- if .Values.serviceLoadBalancer.enabled }}
```
Why do you need a second service? This seems to duplicate the existing one that can already be configured to be of type LoadBalancer if desired, which I think is better anyways.

Commit:
[incubator/logstash] Remove service-lb; Support service annotations.

unguiculus commented:
/ok-to-test

unguiculus commented:
Let's give @christian-roggia some more time to review.

christian-roggia commented:
I apologize but unfortunately I won't be able to review PRs for around a week, I am moving to a new house and my workplace is not ready yet.

I suggest to proceed with the PR anyway if it looks good to @unguiculus, I will test and review everything as soon as I will be able to.

unguiculus commented:
/lgtm

rendhalver commented:
Thanks!
This will give me a nice intro to StatefulSets.
